pkg/preflight/preflight.go

```diff
@@ -0,0 +1,119 @@
+package preflight
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+
+	corev1 "github.com/libopenstorage/operator/pkg/apis/core/v1"
+	coreops "github.com/portworx/sched-ops/k8s/core"
+)
+
+var (
+	instance CheckerOps
+	once     sync.Once
+)
+
+type checker struct {
+	providerName     string
+	distributionName string
+}
+
+// CheckerOps is a list of APIs to do preflight check
+type CheckerOps interface {
+	// ProviderName returns the providerName of the cloud provider
+	ProviderName() string
+
+	// K8sDistributionName returns the k8s distribution providerName
+	K8sDistributionName() string
+
+	// CheckCloudDrivePermission checks if permissions for drive operation is granted
+	CheckCloudDrivePermission(cluster *corev1.StorageCluster) error
+}
+
+func (c checker) ProviderName() string {
+	return c.providerName
+}
+
+func (c checker) K8sDistributionName() string {
+	return c.distributionName
+}
+
+func (c checker) CheckCloudDrivePermission(cluster *corev1.StorageCluster) error {
+	return nil
+}
+
+// Instance returns a singleton instance of preflight check
+func Instance() CheckerOps {
+	once.Do(func() {
+		if instance == nil {
+			instance = checker{}
+		}
+	})
+	return instance
+}
+
+// SetInstance replaces the instance with the provided one. Should be used only for testing purposes.
+func SetInstance(i CheckerOps) {
+	instance = i
+}
+
+// InitPreflightChecker initialize the preflight check instance
+func InitPreflightChecker() error {
+	providerName, err := getCloudProviderName()
+	if err != nil {
+		return err
+	}
+	distributionName, err := getK8sDistributionName()
+	if err != nil {
+		return err
+	}
+
+	c := &checker{
+		providerName:     providerName,
+		distributionName: distributionName,
+	}
+	instance = c
+	if IsEKS() {
+		instance = &aws{
+			checker: *c,
+		}
+	}
+
+	return nil
+}
+
+func getCloudProviderName() (string, error) {
```

Comment:
Ideally pre-flight checker should check with cloudprovider package to find the cloud provider and distribution. It is not the responsibility of the checker to worry about finding that. But it's okay for now.

Reply:
Use to have all checks in cloudprovider until we decided to have a preflight interface. Now I only initial singleton for preflight, then cloudprovider can get everything from it. It's simpler with such initialization in one place. will think about if there's better way to do so.